File: windows/RedLine.Core.Tests/AlertContractTests.cs
using RedLine.Core;

namespace RedLine.Core.Tests;

/// <summary>
/// What a toast is built from, against the same fixture the Swift suite asserts against. The
/// shell posts these words verbatim, so a missing one is a notification with nothing in it.
/// </summary>
public class AlertContractTests
{
    private static AlertBatch Load()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "fixtures", "alert-feed.json");
        var parsed = AlertJson.Parse(File.ReadAllText(path));
        Assert.NotNull(parsed);
        return parsed;
    }

    [Fact]
    public void TheFeedTheWatcherWritesParses()
    {
        var batch = Load();
        Assert.Equal(1, batch.Seq);
        Assert.Equal(new DateTimeOffset(2026, 8, 20, 12, 0, 0, TimeSpan.Zero), batch.At);
        Assert.Equal(2, batch.Events.Count);
    }

    [Fact]
    public void EveryEventCarriesTheWordsToPost()
    {
        foreach (var alert in Load().Events)
        {
            Assert.NotEqual("", alert.Id);
            Assert.NotEqual("", alert.Kind);
            Assert.NotEqual("", alert.Title);
            Assert.NotEqual("", alert.Body);
            Assert.NotEqual("", alert.Provider);
        }
    }

    /// <summary>
    /// A limit reached is the only one that makes a noise, and the engine says which, so a
    /// toast cannot decide to be louder than a notification on macOS.
    /// </summary>
    [Fact]
    public void OnlyTheReachedLimitMakesASound()
    {
        var batch = Load();
        var loud = Assert.Single(batch.Events, e => e.Sound);
        Assert.Equal("limit_reached", loud.Kind);
        Assert.Equal("Codex", loud.Provider);

        var threshold = Assert.Single(batch.Events, e => e.Kind == "threshold");
        Assert.False(threshold.Sound);
        Assert.Equal(80, threshold.Percent);
    }

    /// <summary>Only a threshold has a percentage to report, and a zero would be one.</summary>
    [Fact]
    public void NothingButAThresholdCarriesAPercent()
    {
        Assert.Null(Assert.Single(Load().Events, e => e.Kind == "limit_reached").Percent);
    }
}

File: Tests/RedlineCoreTests/AlertContractTests.swift
// What a shell reads to deliver an alert. The decision is the engine's; the shell only posts,
// so every word it needs has to be in this file.
//
// RedLine.Core.Tests reads the same fixture. Rename a key here and one of the two fails.
import XCTest
@testable import RedlineCore

final class AlertContractTests: XCTestCase {
    private var fixture: URL {
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("windows/RedLine.Core.Tests/fixtures/alert-feed.json")
    }

    /// The events the fixture was published from, so the whole file can be rebuilt and
    /// compared rather than picked at key by key.
    private var events: [AlertEvent] {
        [
            AlertEvent(kind: .threshold(80), provider: "Claude", key: "five_hour",
                       title: "Claude · Session · 5h", body: "80% used, resets at 4:00 PM",
                       id: "Claude:five_hour:1787203800:80"),
            AlertEvent(kind: .limitReached, provider: "Codex", key: "five_hour",
                       title: "Codex · Session · 5h", body: "Limit reached",
                       id: "Codex:five_hour:1787203800:100"),
        ]
    }

    private func recorded() throws -> [String: Any] {
        let json = try JSONSerialization.jsonObject(with: Data(contentsOf: fixture))
        return try XCTUnwrap(json as? [String: Any])
    }

    /// Rebuilt and compared after parsing, not as text: a double prints differently per
    /// platform, and the fixture must not fail on the machine that did not write it.
    func testWhatIsPublishedIsStillWhatTheFixtureHolds() throws {
        let out = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-alert-contract-\(UUID().uuidString).json")
        defer { try? FileManager.default.removeItem(at: out) }

        XCTAssertEqual(AlertFeed.publish(events,
                                         now: Date(timeIntervalSince1970: 1_787_227_200),
                                         to: out), 1)
        let published = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: Data(contentsOf: out)) as? [String: Any])
        XCTAssertEqual(NSDictionary(dictionary: published),
                       NSDictionary(dictionary: try recorded()),
                       "the shape a shell delivers from has moved; regenerate "
                       + "windows/RedLine.Core.Tests/fixtures/alert-feed.json")
    }

    /// A shell posts these words verbatim, so they have to arrive with the event rather than
    /// being assembled from a key and a percentage on the other side.
    func testEveryEventCarriesTheWordsToPost() throws {
        let events = try XCTUnwrap(try recorded()["events"] as? [[String: Any]])
        XCTAssertEqual(events.count, 2)
        for event in events {
            for key in ["id", "kind", "provider", "key", "title", "body", "sound"] {
                XCTAssertNotNil(event[key], "an event with no \(key)")
            }
            XCTAssertFalse((event["title"] as? String ?? "").isEmpty)
            XCTAssertFalse((event["body"] as? String ?? "").isEmpty)
        }
    }

    func testTheReachedLimitIsTheOnlyOneThatMakesASound() throws {
        let events = try XCTUnwrap(try recorded()["events"] as? [[String: Any]])
        let loud = events.filter { $0["sound"] as? Bool == true }
        XCTAssertEqual(loud.count, 1)
        XCTAssertEqual(loud.first?["kind"] as? String, "limit_reached")
    }
}
